test/tls-provider.c

```diff
@@ -748,89 +1288,2265 @@ static const OSSL_DISPATCH xor_keymgmt_functions[] = {
+        OPENSSL_free(pxor_sigctx);
+        pxor_sigctx = NULL;
+        ERR_raise(ERR_LIB_USER, ERR_R_MALLOC_FAILURE);
+    }
+    return pxor_sigctx;
+}
+
+static int xor_sig_setup_md(PROV_XORSIG_CTX *ctx,
+                        const char *mdname, const char *mdprops)
+{
+    EVP_MD *md;
+
+    if (mdprops == NULL)
+        mdprops = ctx->propq;
+
+    md = EVP_MD_fetch(ctx->libctx, mdname, mdprops);
+
+    if ((md == NULL)||(EVP_MD_nid(md)==NID_undef)) {
+        if (md == NULL)
+            ERR_raise_data(ERR_LIB_USER, XORPROV_R_INVALID_DIGEST,
+                           "%s could not be fetched", mdname);
+        EVP_MD_free(md);
+        return 0;
+    }
+
+    EVP_MD_CTX_free(ctx->mdctx);
+    ctx->mdctx = NULL;
+    EVP_MD_free(ctx->md);
+    ctx->md = NULL;
+
+    OPENSSL_free(ctx->aid);
+    ctx->aid = NULL;
+    ctx->aid_len = xor_get_aid(&(ctx->aid), ctx->sig->tls_name);
+
+    ctx->mdctx = NULL;
+    ctx->md = md;
+    OPENSSL_strlcpy(ctx->mdname, mdname, sizeof(ctx->mdname));
+    return 1;
+}
+
+static int xor_sig_signverify_init(void *vpxor_sigctx, void *vxorsig,
+                                   int operation)
+{
+    PROV_XORSIG_CTX *pxor_sigctx = (PROV_XORSIG_CTX *)vpxor_sigctx;
+
+    if ( pxor_sigctx == NULL || vxorsig == NULL)
+        return 0;
+    xor_freekey(pxor_sigctx->sig);
+    if (!xor_key_up_ref(vxorsig)) return 0;
+    pxor_sigctx->sig = vxorsig;
+    pxor_sigctx->operation = operation;
+    if ( (operation==EVP_PKEY_OP_SIGN && !pxor_sigctx->sig) ||
+         (operation==EVP_PKEY_OP_VERIFY && !pxor_sigctx->sig)) {
+        ERR_raise(ERR_LIB_USER, XORPROV_R_INVALID_KEY);
+        return 0;
+    }
+    return 1;
+}
+
+static int xor_sig_sign_init(void *vpxor_sigctx, void *vxorsig,
+                             const OSSL_PARAM params[])
+{
+    return xor_sig_signverify_init(vpxor_sigctx, vxorsig, EVP_PKEY_OP_SIGN);
+}
+
+static int xor_sig_verify_init(void *vpxor_sigctx, void *vxorsig,
+                               const OSSL_PARAM params[])
+{
+    return xor_sig_signverify_init(vpxor_sigctx, vxorsig, EVP_PKEY_OP_VERIFY);
+}
+
+static int xor_sig_sign(void *vpxor_sigctx, unsigned char *sig, size_t *siglen,
+                    size_t sigsize, const unsigned char *tbs, size_t tbslen)
+{
+    PROV_XORSIG_CTX *pxor_sigctx = (PROV_XORSIG_CTX *)vpxor_sigctx;
+    XORKEY* xorkey = pxor_sigctx->sig;
+
+    size_t max_sig_len = EVP_MAX_MD_SIZE;
+    unsigned int xor_sig_len = 0;
+    int rv = 0;
+
+    if (!xorkey || !xorkey->hasprivkey) {
+      ERR_raise(ERR_LIB_USER, XORPROV_R_NO_PRIVATE_KEY);
+      return rv;
+    }
+
+    if (sig == NULL) {
+      *siglen = max_sig_len;
+      return 1;
+    }
+    if (*siglen < max_sig_len) {
+        ERR_raise(ERR_LIB_USER, XORPROV_R_BUFFER_LENGTH_WRONG);
+        return rv;
+    }
+
+    /* create HMAC using XORKEY as key and hash as data:
+     * No real crypto, just for test, don't do this at home!
+     */
```

Comment:
...

Reply:
fixed in https://github.com/openssl/openssl/pull/19312/commits/0330203fc25c8b2d08c052ef8070ab6f25d17209